packages/postgres-database/src/simcore_postgres_database/utils_folders_v2.py

```diff
@@ -0,0 +1,916 @@
+        [
+            folders_access_rights.c.folder_id,
+            folders_access_rights.c.gid,
+            folders_access_rights.c.traversal_parent_id,
+            folders_access_rights.c.original_parent_id,
+            folders_access_rights.c.read,
+            folders_access_rights.c.write,
+            folders_access_rights.c.delete,
+            sa.literal_column("access_rights_cte.level + 1").label("level"),
+        ]
+    ).select_from(
+        folders_access_rights.join(
+            access_rights_cte,
+            folders_access_rights.c.folder_id == access_rights_cte.c.original_parent_id,
+        )
+    )
+
+    # Combine anchor and recursive CTE
+    folder_hierarchy = access_rights_cte.union_all(recursive)
+
+    # Final query to filter and order results
+    query = (
+        sa.select(
+            [
+                folder_hierarchy.c.folder_id,
+                folder_hierarchy.c.gid,
+                folder_hierarchy.c.traversal_parent_id,
+                folder_hierarchy.c.original_parent_id,
+                folder_hierarchy.c.read,
+                folder_hierarchy.c.write,
+                folder_hierarchy.c.delete,
+                folder_hierarchy.c.level,
+            ]
+        )
+        .where(
+            (
+                _get_all_permissions(permissions, folder_hierarchy)
+                if enforece_all_permissions
+                else _get_true_permissions(permissions, folder_hierarchy)
+            )
+            if permissions
+            else True
+        )
+        .where(folder_hierarchy.c.original_parent_id.is_(None))
+        .where(folder_hierarchy.c.gid == gid)
+        .order_by(folder_hierarchy.c.level.asc())
+    )
+
+    result = await connection.execute(query.params(start_folder_id=folder_id))
+    top_most_parent: RowProxy | None = await result.fetchone()
+    return top_most_parent
+
+
+async def _check_folder_and_access(
+    connection: SAConnection,
+    folder_id: _FolderID,
+    gid: _GroupID,
+    *,
+    permissions: _FolderPermissions,
+    enforece_all_permissions: bool,
+) -> RowProxy:
+    """
+    Raises:
+        FolderNotFoundError
+        FolderNotSharedWithGidError
+        InsufficientPermissionsError
+    """
+    folder_entry: int | None = await connection.scalar(
+        sa.select([folders.c.id]).where(folders.c.id == folder_id)
+    )
+    if not folder_entry:
+        raise FolderNotFoundError(folder_id=folder_id)
+
+    # check if folder was shared
+    top_most_access_rights_without_permissions = (
+        await _get_top_most_access_rights_entry(
+            connection,
+            folder_id,
+            gid,
+            permissions=None,
+            enforece_all_permissions=False,
+        )
+    )
+    if not top_most_access_rights_without_permissions:
+        raise FolderNotSharedWithGidError(folder_id=folder_id, gid=gid)
+
+    # check if there are permissions
+    top_most_parent_with_permissions = await _get_top_most_access_rights_entry(
```

Comment:
Well it's done to compute different things. In one situation you compute if the folder was shared with you. In the second call here, you check if it has the correct permissions.
Otherwise you would not be able to distinguish between the two.
And I did not find a way to do it with one query. Since getting no results can't tell me if the entry was not shared or the access rights are wrong.